### PR TITLE
Inject writers in the exporter.

### LINKED ISSUE #269: Make the writers configurable services
My client wants to open their CSV files with Microsoft Excel or LibreOffice, but Excel does not automatically detect that commas are in used and defaults to SSV (Semicolon-separated values).

I think a way to fix that would be to make the CsvWriter configurable services (and the other writers while we're at it).

An ugly workaround is to create another Exporter configured with semicolons, and overwrite the service definition, like this : 

```yml
    sonata.admin.exporter: # ugly service id override
        class: MyProject\AppBundle\Admin\Export\Exporter
```

The DIP violation is here : https://github.com/sonata-project/SonataCoreBundle/blob/3.x/Exporter/Exporter.php#L48

### PR DESCRIPTION
**Pull request by greg0ire**
Closes #269 

### Changelog

```markdown
### Added
- `Sonata\CoreBundle\Exporter\Exporter::__construct()` now accepts a hash of writers.
- `Sonata\CoreBundle\Exporter\Exporter::__addWriter()` allows to add writers through the new `sonata_core.exporter.writer` tag.
- Existing exporter writers can be configured with container parameters.
```

### Todo

- [x] test the service definitions
- [x] document the new service
- [x] introduce container parameters to make the default services configurable
- [x] find a way to let people add new writers
- [x] find a way to let people pick and choose default writers

### Subject

Being able to inject any writer means being able to inject custom writers, but also configure existing writers otherwise. Both are interesting.

### CODE REVIEW

File: Exporter/Exporter.php [hunk 1]
```diff
@@ -14,13 +14,51 @@
...
+    {
+        $this->writers = $writers;
+
+        // NEXT MAJOR : remove this fallback system
+        if ($writers === array()) {
+            $this->writers = array(
+                'csv' => new CsvWriter('php://output', ',', '"', '', true, true),
+                'json' => new JsonWriter('php://output'),
+                'xls' => new XlsWriter('php://output'),
+                'xml' => new XmlWriter('php://output'),
```
Comment (greg0ire) [inline Exporter/Exporter.php]: Maybe `WriterInterface` shoud have a `getFormatName` method?

File: Exporter/Exporter.php [hunk 2]
```diff
@@ -31,35 +69,39 @@ class Exporter
...
+        $writer = $this->writers[$format];
 
         $callback = function () use ($source, $writer) {
             $handler = \Exporter\Handler::create($source, $writer);
             $handler->export();
         };
 
         return new StreamedResponse($callback, 200, array(
-            'Content-Type' => $contentType,
+            'Content-Type' => $this->getContentTypeForWriter($writer),
```
Comment (OskarStark) [inline Exporter/Exporter.php]: what about `getContentType($writer)` ?

Comment (greg0ire) [inline Exporter/Exporter.php]: @OskarStark fixed

File: Tests/Exporter/ExporterTest.php [hunk 3]
```diff
@@ -16,17 +16,21 @@
 
 class ExporterTest extends \PHPUnit_Framework_TestCase
 {
-    /**
-     * @expectedException RuntimeException
-     */
     public function testFilter()
     {
+        $this->setExpectedException('RuntimeException', 'Invalid "foo" format');
```
Comment (Soullivaneuh) [inline Tests/Exporter/ExporterTest.php]: Ninja hack, this is not related to the PR.

File: Exporter/Exporter.php [hunk 4]
```diff
@@ -14,13 +14,51 @@
...
+ */
 class Exporter
 {
     /**
+     * @var array[string => WriterInterface]
+     */
+    private $writers;
+
+    /**
+     * NEXT MAJOR : remove default value.
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: `NEXT_MAJOR:`

File: Exporter/Exporter.php [hunk 5]
```diff
@@ -14,13 +14,51 @@
 use Exporter\Source\SourceIteratorInterface;
 use Exporter\Writer\CsvWriter;
 use Exporter\Writer\JsonWriter;
+use Exporter\Writer\WriterInterface;
 use Exporter\Writer\XlsWriter;
 use Exporter\Writer\XmlWriter;
 use Symfony\Component\HttpFoundation\StreamedResponse;
 
+/**
+ * NEXT MAJOR : make this class final.
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: `NEXT_MAJOR:`

Comment (Soullivaneuh) [inline Exporter/Exporter.php]: And this is not related to the PR i think. Would be better to have an issue for that.

File: Exporter/Exporter.php [hunk 6]
```diff
@@ -14,13 +14,51 @@
...
+    /**
+     * NEXT MAJOR : remove default value.
+     *
+     * @param array[string => WriterInterface] writers a hash of allowed writers
+     */
+    public function __construct(array $writers = array())
+    {
+        $this->writers = $writers;
+
+        // NEXT MAJOR : remove this fallback system
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: * `NEXT_MAJOR:`
* Is this for BC? Could we put some deprecation warning?

File: Tests/Exporter/ExporterTest.php (hunk 3 above)
Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: I slightly improved the error message, and wanted to test that. I can make it a separate commit if necessary, and that will be cumbersome, but a separate PR for that would be too much.

File: Exporter/Exporter.php (hunk 6 above)
Comment (greg0ire) [inline Exporter/Exporter.php]: Yes it is, I'll add a deprecation warning.

Comment (greg0ire) [inline Exporter/Exporter.php]: It forced me to rewrite the tests with the new way, which is a good thing :+1:

File: Tests/Exporter/ExporterTest.php [hunk 7]
```diff
@@ -12,21 +12,30 @@
...
         $source = $this->getMock('Exporter\Source\SourceIteratorInterface');
+        $writer = $this->getMock('Exporter\Writer\WriterInterface');
 
-        $exporter = new Exporter();
```
Comment (OskarStark) [inline Tests/Exporter/ExporterTest.php]: we should keep a test with an empty constructor, tot test if the fallback works, shouldn't we?

Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: You're right, I'll mark it as `@legacy`

Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: @OskarStark done

**Comment by greg0ire** [comment]
Related : https://github.com/sonata-project/exporter/pull/90

**Comment by greg0ire** [comment]
@sonata-project/contributors please review

File: Exporter/Exporter.php [hunk 8]
```diff
@@ -31,35 +71,39 @@ class Exporter
...
+    private function getContentType(WriterInterface $writer)
+    {
+        $mapping = array(
+            'Exporter\Writer\CsvWriter' => 'text/csv',
+            'Exporter\Writer\JsonWriter' => 'application/json',
+            'Exporter\Writer\XlsWriter' => 'application/vnd.ms-excel',
+            'Exporter\Writer\XmlWriter' => 'text/xml',
+        );
+
+        return $mapping[get_class($writer)];
```
Comment (OskarStark) [inline Exporter/Exporter.php]: this could result in an undefined index, couldn't it?

Comment (greg0ire) [inline Exporter/Exporter.php]: Uuuuh you're right! How did I not see this?!

Comment (greg0ire) [inline Exporter/Exporter.php]: @OskarStark fixed

File: Exporter/Exporter.php [hunk 9]
```diff
@@ -14,13 +14,53 @@
...
 {
     /**
+     * @var array[string => WriterInterface]
+     */
+    private $writers;
+
+    /**
+     * NEXT MAJOR: remove default value.
+     *
+     * @param array[string => WriterInterface] writers a hash of allowed writers
```
Comment (core23) [inline Exporter/Exporter.php]: Should be `@param WriterInterace[] ...`

AFAIK your notation is not offical supported and doesn't work with autocompletion.

Comment (greg0ire) [inline Exporter/Exporter.php]: Do you mean it prevents autocompletion ?

Comment (greg0ire) [inline Exporter/Exporter.php]: And then, there is this other notation : http://stackoverflow.com/a/1763425/353612

Comment (greg0ire) [inline Exporter/Exporter.php]: So it should be `WriterInterface[]` ?

Comment (greg0ire) [inline Exporter/Exporter.php]: @core23 fixed

**Comment by OskarStark** [comment]
LGTM 👍 

@Soullivaneuh can you please give us a final review?

**Comment by greg0ire** [comment]
Travis says "offline for maintenance" :S

File: Tests/Exporter/ExporterTest.php [hunk 10]
```diff
@@ -12,21 +12,61 @@
...
     }
 
+    public function testConstructorRejectsNonWriters()
+    {
+        $this->setExpectedException('InvalidArgumentException', 'Writer for format "made-up-format"');
+        new Exporter(array('made-up-format' => 'Not even an object'));
+    }
+
+    /**
+     * NEXT MAJOR: remove this test.
```
Comment (Soullivaneuh) [inline Tests/Exporter/ExporterTest.php]: `NEXT_MAJOR`

https://github.com/sonata-project/SonataCoreBundle/blob/3.x/CONTRIBUTING.md#the-base-branch

File: Exporter/Exporter.php [hunk 11]
```diff
@@ -31,35 +71,39 @@ class Exporter
...
 
-        return new StreamedResponse($callback, 200, array(
-            'Content-Type' => $contentType,
+        $headers = array(
             'Content-Disposition' => sprintf('attachment; filename="%s"', $filename),
-        ));
+        );
+
+        /*
+         * NEXT MAJOR: rely on https://github.com/sonata-project/exporter/pull/90
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Why should this be done on next major?

**Comment by Soullivaneuh** [comment]
@greg0ire Should we wait https://github.com/sonata-project/exporter/pull/90#issuecomment-222369290?

File: Exporter/Exporter.php (hunk 11 above)
Comment (greg0ire) [inline Exporter/Exporter.php]: Because it will imply changing the composer version constraints on the exporter, which is a BC-break

File: Tests/Exporter/ExporterTest.php (hunk 10 above)
Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: @Soullivaneuh fixed

**Comment by greg0ire** [comment]
@Soullivaneuh : why is StyleCI failing ?

**Comment by greg0ire** [comment]
> @greg0ire Should we wait sonata-project/exporter#90 (comment)?

Yes, it might be better, if it is rejected or amended for some reason.

File: Exporter/Exporter.php (hunk 11 above)
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Your PR is for a new minor version of exporter. Changing version constraint for a new minor is not BC break.

File: Exporter/Exporter.php [hunk 12]
```diff
@@ -31,35 +71,39 @@ class Exporter
...
 
-        return new StreamedResponse($callback, 200, array(
-            'Content-Type' => $contentType,
+        $headers = array(
             'Content-Disposition' => sprintf('attachment; filename="%s"', $filename),
-        ));
+        );
+
+        /*
+         * NEXT_MAJOR: rely on https://github.com/sonata-project/exporter/pull/90
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Not sure about that, why on next major?

The `getMimeType` returns in https://github.com/sonata-project/exporter/pull/90 is the same, isn't it?

Comment (greg0ire) [inline Exporter/Exporter.php]: Because it implies bumping the version constraint for exporter

Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Yes, but a minor one, right?

In this case, this is not BC break.

Sullivan SENECHAL

Comment (greg0ire) [inline Exporter/Exporter.php]: > In this case, this is not BC break.

as seen yesterday on twitter you mean ? Right…

Comment (greg0ire) [inline Exporter/Exporter.php]: Well let's wait for sonata-project/exporter#90 then, shall we ?

Comment (Soullivaneuh) [inline Exporter/Exporter.php]: > as seen yesterday on twitter you mean ? Right…

No. Yesterday I was talking about PHP. You know, the gray area...

Here, this is a constraint change for a minor version of a composer package.

Plus, it's a package of ours and we have sure the BC way is kept.

So no, it's not BC break at all.

The same way was adopt for PR concerning class renaming from `Admin` to `AbstractAdmin`

> Well let's wait for sonata-project/exporter#90 then, shall we ?

Yes. :+1:

**Comment by OskarStark** [comment]
@greg0ire what is the status of this PR?

**Comment by greg0ire** [comment]
Blocked by sonata-project/exporter#90, which is blocked by https://github.com/sonata-project/exporter/pull/114 . Please unblock the whole chain guys.

**Comment by greg0ire** [comment]
More precisely : this need to be reworked once sonata-project/exporter#90 is merged, because it will be greatly simplified.

**Comment by Soullivaneuh** [comment]
> More precisely : this need to be reworked once sonata-project/exporter#90 is merged, because it will be greatly simplified.

And I think wait for a new release of exporter, am I right?

**Comment by greg0ire** [comment]
> And I think wait for a new release of exporter, am I right?

Yes, exactly. This MR should touch the Composer manifest. Which by our standards, is a BC-break. Which means I should reopen it on master, closing it for now.

**Comment by Soullivaneuh** [comment]
> Which by our standards, is a BC-break.

Only for important changes like PHP or Symfony.

For the other, except we have to change for a next major of a deps, this is not needed.

Plus, it's an internal project.

**Comment by greg0ire** [comment]
Ok then :)

**Comment by Soullivaneuh** [comment]
@greg0ire You can now play with it: https://github.com/sonata-project/exporter/releases/tag/1.5.0

**Comment by greg0ire** [comment]
Great, will do!

File: Exporter/Exporter.php [hunk 13]
```diff
@@ -14,13 +14,45 @@
...
+     * @var TypedWriterInterface[]
+     */
+    private $writers;
+
+    /**
+     * NEXT_MAJOR: remove default value.
+     *
+     * @param TypedWriterInterface[] $writers an array of allowed typed writers, indexed by format name
+     */
+    public function __construct(array $writers = array())
```
Comment (greg0ire) [inline Exporter/Exporter.php]: It would be a BC-break, no ?

Comment (greg0ire) [inline Exporter/Exporter.php]: ping @Soullivaneuh

Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Oh yes, you are right.

File: Tests/Exporter/ExporterTest.php [hunk 14]
```diff
@@ -12,21 +12,64 @@
...
 
-        $exporter = new Exporter();
+        $exporter = new Exporter(array($writer));
         $exporter->getResponse('foo', 'foo', $source);
     }
 
+    public function testConstructorRejectsNonTypedWriters()
+    {
+        $this->setExpectedException(
+            PHP_VERSION_ID < 70000 ? 'PHPUnit_Framework_Error' : 'TypeError',
```
Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: @Soullivaneuh : I tried detecting `TypeError` here instead, but there is a polyfill for it, which means it will not work.

**Comment by greg0ire** [comment]
@sonata-project/contributors , this is ready for review.

File: Tests/Exporter/ExporterTest.php (hunk 14 above)
Comment (Soullivaneuh) [inline Tests/Exporter/ExporterTest.php]: First of all, why trying to expect a phpunit exception?

Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: It's a feature of phpunit, it converts errors to exception, so that you can assert that an error should occur. But do you mean I should except something less precise, like maybe `\Exception` ?

Comment (Soullivaneuh) [inline Tests/Exporter/ExporterTest.php]: Oh ok, special feature for PHP < 7.0?

Ok I think this is acceptable. But please use `version_compare`, this is more readable.

Comment (greg0ire) [inline Tests/Exporter/ExporterTest.php]: @Soullivaneuh done

File: Exporter/Exporter.php [hunk 15]
```diff
@@ -31,35 +63,36 @@ class Exporter
...
+    }
+
+    /**
+     * The main benefit of this method is the type hinting.
+     *
+     * @param TypedWriterInterface $writer a possible writer for exporting data
+     */
+    final private function addWriter(TypedWriterInterface $writer)
+    {
+        $this->writers[$writer->getFormat()] = $writer;
```
Comment (Soullivaneuh) [inline Exporter/Exporter.php]: Why not just get an associative array on the constructor?

**Comment by Soullivaneuh** [comment]
`Exporter` is a service. Nothing to update about this?

**Comment by Soullivaneuh** [comment]
**PLEASE DON'T MERGE IT FOR NOW**

I found a nasty issue about exporter service. I'll dig on it.

**Comment by Soullivaneuh** [comment]
Wait https://github.com/sonata-project/SonataCoreBundle/pull/308 and a patch release before merge.

File: Exporter/Exporter.php (hunk 15 above)
Comment (greg0ire) [inline Exporter/Exporter.php]: Because that way you benefit from the type hinting (that's what I tried to explain in the method's documentation)

**Comment by Soullivaneuh** [comment]
@greg0ire Could you please rebase and update your PR?

**Comment by greg0ire** [comment]
@Soullivaneuh done

**Comment by greg0ire** [comment]
But it needs some tests, and I think I will add container parameters for the csv writer in this PR (or do you feel otherwise?)

File: Resources/config/exporter.xml [hunk 16]
```diff
@@ -1,6 +1,30 @@
 <?xml version="1.0" encoding="UTF-8"?>
 <container xmlns="http://symfony.com/schema/dic/services" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://symfony.com/schema/dic/services http://symfony.com/schema/dic/services/services-1.0.xsd">
     <services>
-        <service id="sonata.core.exporter" class="Sonata\CoreBundle\Exporter\Exporter"/>
+        <service id="sonata.core.exporter.writer.csv" class="Exporter\Writer\CsvWriter" public="false">
+            <argument>php://output</argument>
+            <argument>,</argument>
```
Comment (Soullivaneuh) [inline Resources/config/exporter.xml]: Hmm does not `CsvWriter` have default values for that? Maybe same for the first argument.

BTW, you made the services private. But what if the user want to change this?

File: Resources/config/exporter.xml [hunk 17]
```diff
@@ -1,6 +1,30 @@
...
+        </service>
+        <service id="sonata.core.exporter.writer.xls" class="Exporter\Writer\XlsWriter" public="false">
+            <argument>php://output</argument>
+        </service>
+        <service id="sonata.core.exporter.writer.xml" class="Exporter\Writer\XmlWriter" public="false">
+            <argument>php://output</argument>
+        </service>
+        <service id="sonata.core.exporter" class="Sonata\CoreBundle\Exporter\Exporter">
+            <argument type="collection">
+                <argument type="service" id="sonata.core.exporter.writer.csv" />
```
Comment (Soullivaneuh) [inline Resources/config/exporter.xml]: Same here. You pass an array of service. Good.

But if I want to add one writer on it without dropping the default ones?

This is maybe for this @rande talked about factories...

File: Resources/config/exporter.xml (hunk 16 above)
Comment (greg0ire) [inline Resources/config/exporter.xml]: This is still in progress. I'll keep them private and use container parameters instead

File: Resources/config/exporter.xml (hunk 17 above)
Comment (greg0ire) [inline Resources/config/exporter.xml]: I think this should be solved with tagged services

Comment (greg0ire) [inline Resources/config/exporter.xml]: Also, maybe it is already possible to add a method call to `addWriter` (not sure if this works for private services)

**Comment by greg0ire** [comment]
I ended a with the following : 

1. If the exporter is called without parameters, the `$writers` defaults to null, and this is the legacy case, to be removed on next major.
2. If the exporter is called with an array, add all the writers in the array
3. To avoid the deprecation notice introduced by 1 but still be BC, I inject an empty array, and then, in the extension, I tag services that are part of the list of default writers.

**Comment by OskarStark** [comment]
regarding the CHANGELOG:

what about:
```diff
- The constructor method of `Sonata\CoreBundle\Exporter\Exporter` now accepts a hash of writers
+ `Sonata\CoreBundle\Exporter\Exporter::__construct()` now accepts a hash of writers
```

**Comment by greg0ire** [comment]
Good idea, I'll use that instead!

File: DependencyInjection/Compiler/ExporterCompilerPass.php [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Class StatusRendererCompilerPass.
+ *
+ *
+ * @author Grégoire Paris <postmaster@greg0ire.fr>
+ */
+class ExporterCompilerPass implements CompilerPassInterface
```
Comment (core23) [inline DependencyInjection/Compiler/ExporterCompilerPass.php]: Final?

File: DependencyInjection/Compiler/ExporterCompilerPass.php [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+ */
+
+namespace Sonata\CoreBundle\DependencyInjection\Compiler;
+
+use Symfony\Component\DependencyInjection\Compiler\CompilerPassInterface;
+use Symfony\Component\DependencyInjection\ContainerBuilder;
+use Symfony\Component\DependencyInjection\Reference;
+
+/**
+ * Class StatusRendererCompilerPass.
```
Comment (core23) [inline DependencyInjection/Compiler/ExporterCompilerPass.php]: We don't need this

File: Exporter/Exporter.php [hunk 20]
```diff
@@ -31,35 +65,36 @@ class Exporter
...
+
+        return new StreamedResponse($callback, 200, $headers);
+    }
+
+    /**
+     * The main benefit of this method is the type hinting.
+     *
+     * @param TypedWriterInterface $writer a possible writer for exporting data
+     */
+    final private function addWriter(TypedWriterInterface $writer)
```
Comment (core23) [inline Exporter/Exporter.php]: Why final? Nobody can extend this method.

Comment (Soullivaneuh) [inline Exporter/Exporter.php]: On some tricky way, they actually can.

Comment (core23) [inline Exporter/Exporter.php]: Started a new discussion to keep this PR clean.

https://github.com/sonata-project/dev-kit/issues/185

File: DependencyInjection/Compiler/ExporterCompilerPass.php (hunk 19 above)
Comment (greg0ire) [inline DependencyInjection/Compiler/ExporterCompilerPass.php]: Plus, it's wrong :P

File: Exporter/Exporter.php (hunk 20 above)
Comment (greg0ire) [inline Exporter/Exporter.php]: Fixed

File: DependencyInjection/SonataCoreExtension.php [hunk 21]
```diff
@@ -174,4 +176,15 @@ protected function deprecateSlugify(ContainerBuilder $container)
             $definition->setDeprecated(true);
         }
     }
+
+    final private function configureExporter(ContainerBuilder $container, array $config)
```
Comment (core23) [inline DependencyInjection/SonataCoreExtension.php]: -final

Comment (greg0ire) [inline DependencyInjection/SonataCoreExtension.php]: Done, sorry… :sweat_smile:

**Comment by OskarStark** [comment]
LGTM 👍

**Comment by greg0ire** [comment]
ping @sonata-project/contributors : do we merge this here first, then move it to the exporter library, or should I do this inside the library directly?

**Comment by Soullivaneuh** [comment]
If we move it to exporter after, this mean deprecated fresh functionalities, right?

In this case, the second option looks better.

**Comment by greg0ire** [comment]
> If we move it to exporter after, this mean deprecated fresh functionalities, right?

Right. I'll rework this exporter-side.
